src/track/trackref.h

```diff
@@ -0,0 +1,146 @@
+    }
+    static QString toCanonicalLocation(const QFileInfo& fileInfo) {
+        return fileInfo.canonicalFilePath();
+    }
+    static QString toDirectory(const QFileInfo& fileInfo) {
+        return fileInfo.absolutePath();
+    }
+    static QString toFileName(const QFileInfo& fileInfo) {
+        return fileInfo.fileName();
+    }
+    static quint64 toFileSize(const QFileInfo& fileInfo) {
+        return fileInfo.size();
+    }
+    static QDateTime toFileCreatedAt(const QFileInfo& fileInfo) {
+        return fileInfo.created();
+    }
+    static QDateTime toFileLastModifiedAt(const QFileInfo& fileInfo) {
+        return fileInfo.lastModified();
+    }
+
+    TrackRef() {
+        DEBUG_ASSERT(verifyConsistency());
+    }
+    // Conversion from QFileInfo
+    explicit TrackRef(
+            const QFileInfo& fileInfo,
+            TrackId id = TrackId())
+        : m_location(toLocation(fileInfo)),
+          m_canonicalLocation(toCanonicalLocation(fileInfo)),
+          m_id(std::move(id)) {
+        DEBUG_ASSERT(verifyConsistency());
+    }
+    // Create a copy of an existing TrackRef and set the TrackId.
+    // This is equivalent to invoking the regular copy constructor
+    // followed by a call to setId().
+    TrackRef(
+            const TrackRef& other,
+            TrackId id)
+        : m_location(other.m_location),
+          m_canonicalLocation(other.m_canonicalLocation),
+          m_id(std::move(id)) {
+        DEBUG_ASSERT(!other.hasId() || (other.getId() == id));
+        DEBUG_ASSERT(verifyConsistency());
+    }
+
+    // The human-readable identifier of a track in Mixxx. The location is
+    // immutable and the starting point for accessing a track's file.
+    const QString& getLocation() const {
+        return m_location;
+    }
+    bool hasLocation() const {
+        return !getLocation().isEmpty();
+    }
+
+    // The unique identifier of a track's file at runtime and used
+    // for caching. The canonical location is empty for inexistent
+    // files. Like the id it can only be set once.
+    const QString& getCanonicalLocation() const {
+        return m_canonicalLocation;
+    }
+    bool hasCanonicalLocation() const {
+        return !getCanonicalLocation().isEmpty();
+    }
+    void setCanonicalLocation(QString canonicalLocation) {
+        DEBUG_ASSERT(!hasCanonicalLocation() ||
+                (getCanonicalLocation() == getCanonicalLocation()));
+        m_canonicalLocation = std::move(canonicalLocation);
+        DEBUG_ASSERT(verifyConsistency());
+    }
+
+    // The primary key of a track in the Mixxx library. The id must only
+    // be set once after inserting into or after loading from the database.
+    // Tracks that have not been stored in the database don't have an id.
+    const TrackId& getId() const {
+        return m_id;
+    }
+    bool hasId() const {
+        return getId().isValid();
+    }
+    void setId(TrackId id) {
+        DEBUG_ASSERT(!hasId() || (getId() == id));
```

Comment:
The ID should be set only once, but setting the same ID again is acceptable, 
too.

On 09/17/2015 09:06 AM, Daniel Schürmann wrote:
>
> In src/track/trackref.h 
> <https://github.com/mixxxdj/mixxx/pull/709#discussion_r39716216>:
>
> > +                (getCanonicalLocation() == getCanonicalLocation()));
> > +        m_canonicalLocation = std::move(canonicalLocation);
> > +        DEBUG_ASSERT(verifyConsistency());
> > +    }
> > +
> > +    // The primary key of a track in the Mixxx library. The id must only
> > +    // be set once after inserting into or after loading from the database.
> > +    // Tracks that have not been stored in the database don't have an id.
> > +    const TrackId& getId() const {
> > +        return m_id;
> > +    }
> > +    bool hasId() const {
> > +        return getId().isValid();
> > +    }
> > +    void setId(TrackId id) {
> > +        DEBUG_ASSERT(!hasId() || (getId() == id));
>
> Why we need "|| (getId() == id)"?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/709/files#r39716216>.
>